Bazel rules for the graphs backend.

maemre opened the pull request:
This PR adds
- a target (`:p4c_graphs`) to build the compiler binary with the graphs
  backend, and
- a build rule `p4_graphs` to invoke this backend and merge the graph
  files

so that the graphs backend can be used in other projects using Bazel.

We need a new build rule (rather than specifying a different binary for
`_p4c` in the `p4_library` rule) because the graphs backend does
not use the normal output file arguments and uses a separate argument to
specify the output *directory* to write the graphs in. The names of the
output files depend on the content of the input p4 file, and blaze
outputs cannot be directories (they can be file groups but we do not
know the exact file names without effectively running the p4 compiler).
So, we cannot directly use the output files in a blaze target. Instead,
this build rule concatenates the output graphs into a single GraphViz
file. The output file can later be processed by `gvpack` and `dot` to
produce a single large graph (the original graphs are embedded as
subgraphs). It can also be parsed by other tools to extract specific
graphs.

maemre commented:
@smolkaj can you review this PR?

smolkaj commented on bazel/p4_library.bzl:
```diff
@@ -120,3 +120,103 @@ p4_library = rule(
     toolchains = ["@bazel_tools//tools/cpp:toolchain_type"],
 )
+
+def _p4_graphs_impl(ctx):
```
It looks like this is duplicating a lot of code from _p4_library_impl. Could we reuse the code somehow?

maemre replied:
I added two helpers to reuse argument extraction and shell/cc environment setup

smolkaj commented on bazel/p4_library.bzl:
```diff
@@ -120,3 +120,103 @@ p4_library = rule(
+            allow_files = [".p4", ".h"],
+            default = [],
+        ),
+        "out_file": attr.output(
```
Nit: just "out" would be a bit more idiomatic.

maemre replied:
Done.

smolkaj commented on bazel/p4_library.bzl:
```diff
@@ -120,3 +120,103 @@ p4_library = rule(
+            default = [],
+        ),
+        "out_file": attr.output(
+            doc = "The name of the output DOT file",
```
Nit: a check that the file actually ends in ".dot" would make this rule easier to use.
You can use `str.endswith(str) -> bool` and `fail(str)`.

maemre replied:
Done.

smolkaj reviewed (changes requested):
Looks good to me at a high level.

Could you also extend bazel/example with the new rule? This will make it both easier to use for others. It will also cause the rule to be executed as part of continuous integration, so it will serve as a basic safety net ensuring the rule is at least not completely broken.

smolkaj commented on bazel/p4_library.bzl:
```diff
@@ -120,3 +120,103 @@ p4_library = rule(
+            # Run the compiler
+            "{p4c}" {p4c_args}
+            # Merge all output graphs
+            cat "{graph_dir}"/* > {output_file}
```
Nit: "{graph_dir}"/* -> "{graph_dir}/*"
(unless this brakes the * somehow)

maemre replied:
That breaks the shell globbing by *, I added a comment to explain it.

smolkaj replied:
Thanks for adding the coment!

maemre commented:
@smolkaj I made the changes pointed in the code review. Can you review it again?

The two big changes are:

- I split off two common functionality to separate functions: argument extraction, and running commands in the required shell environment.
- I added a rule to `bazel/example/p4/BUILD` to show how to use the `p4_graphs` rule and to test it.

smolkaj commented on bazel/p4_library.bzl:
```diff
@@ -2,26 +2,62 @@
+
+    return (args, inputs)
+
+def _run_p4c_with_cc(ctx, p4c, command, **run_shell_kwargs):
```
How about simplifying this to "_run_shell_cmd_with_p4c(ctx, command)"? In particular:
- The function can assert that `ctx.executable._p4c` is defined.
- No need to mention cc, that's arguably an implementation detail.
- We can add `run_shell_kwargs` if we need it in the future, currently it is unused.

maemre replied:
The build rule implementations pass their inputs, outputs, and progress messages via `run_shell_kwargs`. I can add them as explicit parameters instead, if that makes sense.

I simplifed it with other changes. I used `if` and `fail` because Starlark does not have assertions (outside unit tests, which do not fail immediately) as far as I am aware of.

smolkaj commented on bazel/p4_library.bzl:
```diff
@@ -2,26 +2,62 @@
 
-def _p4_library_impl(ctx):
-    p4c = ctx.executable._p4c
+def _extract_common_p4c_args(ctx):
```
Could we make this function only do one thing, i.e. extract the common p4c args, as the name suggests?
I think that would make things easier to understand, and extracting the inputs is just a one-liner that we can either duplicate or make a separate function.

maemre replied:
I separated extracting the inputs to a helper function `_extract_p4c_inputs` that is called by the build rule implementations

smolkaj reviewed (changes requested):
Thanks for the changes. Comments inline.

smolkaj commented on bazel/p4_library.bzl:
```diff
@@ -2,27 +2,73 @@
+
+def _extract_p4c_inputs(ctx):
+    """Extract input p4 files to give to p4c from the build rule context."""
+    p4file = ctx.file.src
```
Optionally: consider making this a one-liner.

maemre replied:
I was wary of doing it because of readability but it looks readable as a one-liner, so I made it a one-liner.

smolkaj commented on bazel/p4_library.bzl:
```diff
@@ -2,27 +2,73 @@
+
+def _run_shell_cmd_with_p4c(ctx, command, **run_shell_kwargs):
+    """Run given sequence of shell commands using `run_shell` action after
+setting up the C compiler toolchain.
```
Nit: please indent.

maemre replied:
done

smolkaj commented on bazel/p4_library.bzl:
```diff
@@ -2,27 +2,73 @@
+    return p4deps + [p4file]
+
+def _run_shell_cmd_with_p4c(ctx, command, **run_shell_kwargs):
+    """Run given sequence of shell commands using `run_shell` action after
```
"Runs the given shell command" (since it is only a single string, rather than a list of strings)

maemre replied:
done

smolkaj reviewed (approved):
Some final minor comments. Otherwise LGTM to me.

@mbudiu-vmw, could you take a look when you get a chance?

maemre commented:
@mbudiu-vmw I just wanted to ping you, in case this PR has fallen through the cracks.

mbudiu-vmw commented:
I really know nothing about Bazel, so I cannot review this. But if @smolkaj approved them they are good for me.

smolkaj commented:
Confirming that the changes look good from the technical (Bazel) perspective.
